### PR TITLE
Non-persistent statement logging

### PR DESCRIPTION
**Pull request by umanwizard**
This PR causes a log of statements to be maintained but not persisted, in the tables `mz_prepared_statement_history` and `mz_statement_execution_history`. Logging is controlled by the session variable `statement_logging_sample_rate`, constrained by the system variable `statement_logging_max_sample_rate`.

It works by emitting a "begin" event when the statement begins executing, and an "end" event when it ends. What it means for statement execution to "end" is determined by the type of statement. For most statements, it is the moment when control of the session is given back to the protocol layer. However, for some other statements (notably non-constant `SELECT`s, and `SUBSCRIBE`s), a more complex notion is needed; for example, a `SELECT` is only considered finished after the cluster has returned results, whereas a non-constant `SUBSCRIBE` is never finished unless it errors or the client cancels it.

Thus, we need to pipe the statement logging ID through the various commands and responses that implement these more complex protocols. This is done by adding that ID to the `ExecuteContextExtra` struct. We also make that struct log an error if it's dropped anywhere other than in the coordinator code that handles finishing logging, to ensure that we don't forget somewhere to end a statement.

A follow-up PR will persist the log in the stash, which we may later decide to replace with storing it directly in Persist. In this PR the log is in a built-in table, which is cleared on restart.

### CODE REVIEW

**Review by ParkMyCar — COMMENTED** [review]
Just briefly skimmed this, will take a more in-depth look soon. I'm going to defer accepting to @mjibson or @jkosh44 though

File: src/adapter/src/coord/statement_logging.rs [hunk 1]
```diff
@@ -43,13 +46,116 @@ pub enum PreparedStatementLoggingInfo {
...
+    /// (because it was not sampled). Requiring the opaque `StatementLoggingId` type,
+    /// which is only instantiated by `begin_statement_execution` if the statement is actually logged,
+    /// should prevent this.
+    pub fn end_statement_execution(
+        &mut self,
+        StatementLoggingId(id): StatementLoggingId,
+        reason: StatementEndedExecutionReason,
     ) {
+        let now = self.now_datetime();
+        let now_millis = now.timestamp_millis().try_into().expect("sane system time");
```
Comment (ParkMyCar) [inline src/adapter/src/coord/statement_logging.rs:122]: Instead of taking a timestamp here should we move this into `ExecuteContext::retire` and add it to `Message::RetireExecute`? Since `Message::RetireExecute` is put into a channel and processed sometime after the fact these timestamps might not be very accurate to the time at which retire was called

File: src/adapter/src/coord/statement_logging.rs [hunk 2]
```diff
@@ -43,13 +46,116 @@ pub enum PreparedStatementLoggingInfo {
...
+            ended_at: now_millis,
+        };
+
+        let began_record = self.statement_logging_executions_begun.remove(&id).expect(
+            "matched `begin_statement_execution` and `end_statement_execution` invocations",
+        );
+        let updates = self
+            .catalog
+            .state()
+            .pack_statement_ended_execution_updates(&began_record, &ended_record);
```
Comment (ParkMyCar) [inline src/adapter/src/coord/statement_logging.rs:135]: From a quick look it seems like the only part of this that needs the `Catalog` is a call to `Catalog::resolve_builtin_table(...)`. Can we do that call separately to decouple the rest of the packing process from the catalog? If need be this would allow us to move the work off the coord thread sometime in the future

**Review by ParkMyCar — COMMENTED** [review]
Wooo, this is looking good!

Something I am getting hung up on is needing to "retire" the `ExecutionContext`, part of it was I didn't realize retire == complete == send_response. Maybe renaming this could help, although it could also be my own shortcoming.

File: src/adapter/src/command.rs [hunk 3]
```diff
@@ -584,6 +600,13 @@ pub enum ExecuteResponse {
         #[derivative(Debug = "ignore")]
         span: tracing::Span,
     },
+    /// Like `SendingRows`, but the rows are known to be available
+    /// immediately, and thus the execution is considered ended in the coordinator.
+    SendingRowsImmediate {
+        rows: Vec<Row>,
+        #[derivative(Debug = "ignore")]
+        span: tracing::Span,
+    },
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/adapter/src/command.rs:609]: The key difference between `SendingRowsImmediate` and `SendingRows` is that `SendingRowsImmediate` is considered to have finished in the coordinator (and thus we can retire the statement immediately), whereas `SendingRows` is considered to have finished once the result set becomes available (by which time we have already given control of the session back to the pgwire layer, and so we need to send a message containing the statement UUID back to the coordinator).

File: src/adapter/src/coord/statement_logging.rs [hunk 4]
```diff
@@ -43,13 +46,116 @@ pub enum PreparedStatementLoggingInfo {
...
+    /// passed to `end_statement_execution` to record when it ends.
+    pub fn begin_statement_execution(
+        &mut self,
+        session: &mut Session,
+        params: Params,
+        logging: &Arc<QCell<PreparedStatementLoggingInfo>>,
+    ) -> Option<StatementLoggingId> {
+        let sample_rate = self.statement_execution_sample_rate(session);
+
+        let distribution = Bernoulli::new(sample_rate).expect("rate must be in range [0, 1]");
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/adapter/src/coord/statement_logging.rs:149]: We do enforce that the var is between 0 and 1, [here](https://github.com/MaterializeInc/materialize/blob/fb4c4ebb75/src/sql/src/session/vars.rs#L1405-L1408). I can add a test for that.

File: src/adapter/src/coord.rs [hunk 5]
```diff
@@ -867,6 +1011,24 @@ pub struct Coordinator {
...
+
+    /// Information about sessions that have been started, but which
+    /// have not yet been logged in `mz_session_history`.
+    /// They may be logged as part of a statement being executed (and chosen for logging).
+    statement_logging_unlogged_sessions: BTreeMap<Uuid, SessionHistoryEvent>,
+
+    /// A reproducible RNG for deciding whether to sample statement executions.
+    /// Only used by tests; otherwise, `rand::thread_rng()` is used.
+    /// Controlled by the system var `statement_logging_use_reproducible_rng`.
+    statement_logging_reproducible_rng: rand_chacha::ChaCha8Rng,
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/adapter/src/coord.rs:1033]: You're right, that could make things cleaner.

File: src/adapter/src/coord.rs [hunk 6]
```diff
@@ -867,6 +1011,24 @@ pub struct Coordinator {
...
+    ///
+    /// This map needs to have enough state left over to later retract
+    /// the system table entries (so that we can update them when the
+    /// execution finished.)
+    statement_logging_executions_begun: BTreeMap<Uuid, StatementBeganExecutionRecord>,
+
+    /// Information about sessions that have been started, but which
+    /// have not yet been logged in `mz_session_history`.
+    /// They may be logged as part of a statement being executed (and chosen for logging).
+    statement_logging_unlogged_sessions: BTreeMap<Uuid, SessionHistoryEvent>,
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/adapter/src/coord.rs:1028]: It's a bit hard to imagine these growing too huge without actively malicious behavior (E.g., executing billions of statements that never terminate).

The `statement_logging_execution_begun` entries are removed whenever the statement finishes (things could pile up if there's a bug where we forgot to retire some statement, but the `soft_assert_or_log` in `ExecuteContextExtra` is meant to make that harder). The `statement_logging_unlogged_sessions` stick around until the next reboot (or until the session issues a statement that is logged), but given that by far our biggest user of sessions uses at most a few thousand per day it will only grow to ~15000 max between maintenance windows.

One thing I can do make the risk even less is to drop sessions from this map when they disconnect, because there's no good reason not to.

File: src/adapter/src/coord/statement_logging.rs (hunk 1 above)
Comment (umanwizard) [reply src/adapter/src/coord/statement_logging.rs:122]: That's a good point. I suppose it's ambiguous -- has a statement "finished" once control is returned to the user, or once the coordinator has cleaned up the state related to it? I'm leaning towards implementing your suggestion, though I don't think it's 100% clear.

File: src/adapter/src/coord/statement_logging.rs (hunk 2 above)
Comment (umanwizard) [reply src/adapter/src/coord/statement_logging.rs:135]: I could, but I'm a bit confused -- isn't having the `pack_xyz` methods be on `Catalog` (and internally call `resolve_builtin_table`) the standard pattern? Do we have plans to change that, or is there something specific about this case that means it makes sense to do it elsewhere?

File: src/adapter/src/client.rs [hunk 7]
```diff
@@ -485,6 +488,19 @@ impl SessionClient {
...
 
+    /// Tells the coordinator a statement has finished execution, in the cases
+    /// where we have no other reason to communicate with the coordinator.
+    pub fn retire_execute(
+        &mut self,
+        data: ExecuteContextExtra,
+        reason: StatementEndedExecutionReason,
+    ) {
+        if !data.is_trivial() {
+            let cmd = Command::RetireExecute { data, reason };
```
Comment (mjibson) [inline src/adapter/src/client.rs:499]: I logically find it weird that in the trivial case, `reason` is ignored. Is that ok?

File: src/adapter/src/command.rs [hunk 8]
```diff
@@ -803,14 +830,20 @@ impl ExecuteResponse {
...
+                SendingRows,
+                SendingRowsImmediate,
+                Updated,
+            ],
             PlanKind::Fetch => vec![ExecuteResponseKind::Fetch],
             GrantPrivileges => vec![GrantedPrivilege],
             GrantRole => vec![GrantedRole],
             CopyRows => vec![Inserted],
-            Insert => vec![Inserted, SendingRows],
+            Insert => vec![Inserted, SendingRows, SendingRowsImmediate],
```
Comment (mjibson) [inline src/adapter/src/command.rs:846]:
```suggestion
            Insert => vec![Inserted, SendingRowsImmediate],
```

Pretty sure this will always be the case? If I'm wrong tests will panic.

File: src/adapter/src/command.rs [hunk 9]
```diff
@@ -821,7 +854,7 @@ impl ExecuteResponse {
             }
             PlanKind::Subscribe => vec![Subscribing, CopyTo],
             StartTransaction => vec![StartedTransaction],
-            SideEffectingFunc => vec![SendingRows],
+            SideEffectingFunc => vec![SendingRows, SendingRowsImmediate],
```
Comment (mjibson) [inline src/adapter/src/command.rs:857]:
```suggestion
            SideEffectingFunc => vec![SendingRowsImmediate],
```

Same, should never involve clusterd.

File: src/adapter/src/coord.rs [hunk 10]
```diff
@@ -737,11 +772,122 @@ impl ExecuteContext {
             session,
             extra,
         } = self;
+        let reason = if extra.is_trivial() {
+            None
+        } else {
+            Some(match &result {
+                Ok(ok) => match ok {
+                    ExecuteResponse::CopyTo { resp, .. } => match resp.as_ref() {
+                        // NB [btv]: It's not clear that this combination
```
Comment (mjibson) [inline src/adapter/src/coord.rs:781]: `COPY (SELECT 1) TO STDOUT`? Although I'm confused why the other sending types panic at all, and also why this variant doesn't.

File: src/adapter/src/coord/command_handler.rs [hunk 11]
```diff
@@ -372,43 +369,81 @@ impl Coordinator {
...
+        &mut self,
+        portal_name: String,
+        mut session: Session,
+        tx: ClientTransmitter<ExecuteResponse>,
+        // If this command was part of another execute command
+        // (for example, executing a `FETCH` statement causes an execute to be
+        //  issued for the cursor it references),
+        // then `outer_context` should be `Some`.
+        // This instructs the coordinator that the
+        // outer execute should be considered finished once the inner one is.
```
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:382]: Does this correctly handle things like:

```
BEGIN;
DECLARE c FOR VALUES (1), (2);
FETCH 1 c;
FETCH 1 c;
FETCH 1 c;
COMMIT;
```

Where the first `FETCH` only returns a partial result (even though it is fully executed)?

Also note this exact same sequence can happen by a driver, pgwire sending `Execute` messages with `max_rows`: https://github.com/MaterializeInc/materialize/blob/b636580062bbce260c04a665e613c6dbbdd7b624/test/pgtest/portals.pt#L8

File: src/adapter/src/coord/statement_logging.rs (hunk 2 above)
Comment (mjibson) [reply src/adapter/src/coord/statement_logging.rs:135]: The coordinator can already hand out cheap Catalog references that can be moved off thread if needed. We don't need to change the pack pattern to accommodate.

File: src/pgwire/src/protocol.rs [hunk 12]
```diff
@@ -973,18 +985,30 @@ where
...
+                            outer_ctx_extra,
+                            StatementEndedExecutionReason::Errored { error: msg.clone() },
+                        );
+                    }
                     return self
-                        .error(ErrorResponse::error(
-                            SqlState::INVALID_CURSOR_NAME,
-                            format!("portal {} does not exist", portal_name.quoted()),
-                        ))
+                        .error(ErrorResponse::error(SqlState::INVALID_CURSOR_NAME, msg))
```
Comment (mjibson) [inline src/pgwire/src/protocol.rs:996]: `self.error` is called in a buncha places. Can we pass it `outer_ctx_extra` so it can do the retire logic? Are there any calls to `error` that don't want that to happen?

File: src/pgwire/src/protocol.rs [hunk 13]
```diff
@@ -1390,7 +1456,7 @@ where
                 }
                 command_complete!()
             }
-            ExecuteResponse::Subscribing { rx } => {
+            ExecuteResponse::Subscribing { rx, ctx_extra } => {
```
Comment (mjibson) [inline src/pgwire/src/protocol.rs:1459]: Maybe not for this PR: having `SUBSCRIBE` only log on completion might be bad because in the happy path, it won't complete. Can we make it special so that we log on start so we get full `SUBSCRIBE` text even if a query didn't complete?

I alllmost want the same thing for slow path peeks, because those have a higher potential to crash loop a clusterd, and in those cases we won't have the crash-inducing query.

Comment (umanwizard) [reply src/pgwire/src/protocol.rs:1459]: We will get the sql text in those cases as we log that on start! The main things we log on end are the status and the time to completion. So if a subscribe or slow path select is still running we’ll just show null for those.

File: src/adapter/src/coord/command_handler.rs (hunk 11 above)
Comment (umanwizard) [reply src/adapter/src/coord/command_handler.rs:382]: No, it panics. Good catch.

**Comment by philip-stoev** [comment]
@umanwizard The QA team would like to test this -- do you mind giving us a couple of days?

File: src/pgwire/src/protocol.rs (hunk 12 above)
Comment (umanwizard) [reply src/pgwire/src/protocol.rs:996]: It would be a bit annoying because we call that from tons of places that aren't in the context of a statement execution. So we'd have to pass `None` in all those places.

File: src/adapter/src/client.rs (hunk 7 above)
Comment (umanwizard) [reply src/adapter/src/client.rs:501]: Yeah, because "trivial" here means "there is nothing to log" (e.g., we are not doing statement logging at all for this statement)

**Review by mjibson — APPROVED** [review]

**Comment by philip-stoev** [comment]
Item No 1.  The SQLsmith job from Nightly reports repeated OOMs. This is not the case with the "main" branch. Could it be because a lot of distinct SQL statements are being executed?

**Comment by philip-stoev** [comment]
Item No 2. Should this error out -- attempting to set the session variable to a value higher than the system variable:

```
materialize=> show statement_logging_max_sample_rate;
 statement_logging_max_sample_rate 
-----------------------------------
 0
(1 row)

materialize=> set statement_logging_sample_rate=0.4;
SET
```

**Comment by philip-stoev** [comment]
Item No 3. `statement_logging_sample_rate` is zero by default, so unless the session self-enables logging, no logging will take place. Is this by design? If so, it would not be possible to begin logging queries in a cluster unless the session, that is, the application, cooperates to run the `SET statement_logging_sample_rate` statement for us.

**Comment by philip-stoev** [comment]
Item No 4. I am unable to get parameters to be recorded. I set the global sample rate to `1` and then run:

```
create table t1 (f1 integer);
set statement_logging_sample_rate=1;
prepare q1 as select * from t1 where f1 = $1;
execute q1(1234);
```

but the `params` column in `mz_internal.mz_statement_execution_history` contains just `{}` and no reference to `1234`.

**Comment by philip-stoev** [comment]
Item No 5. Do we also want to record the `HINT` that was emitted as part of the error. For a SQL session, things look like this:

```
ERROR:  canceling statement due to statement timeout
HINT:  Consider increasing the maximum allowed statement duration for this session by setting the statement_timeout session variable. For example, `SET statement_timeout = '60s'`.
```

so the user can use the hint to take an action. But, in the statement logging table, we only record:

```
canceling statement due to statement timeout 
```

so, the user now needs to have another way to get to the hint, either via the UI or the documentation.

**Comment by philip-stoev** [comment]
Item No 6. The `name` column in `mz_internal.mz_prepared_statement_history` is always empty, even if I do:

```
prepare q1 as select * from t1 where f1 = $1;
```

the output looks like this:

```
 f21800ee-f6f6-4d8d-915e-1a9dff92b990 | 879045d6-2b76-4fed-a539-b780e8cd074d |      | prepare q1 as select * from t1 where f1 = $1                    | 2023-08-08 09:25:22.432+00
```

the column before the sql is the `name` column, and it is empty.

**Comment by philip-stoev** [comment]
Item No 7:

The `mz_internal.mz_prepared_statement_history` does not show the database and schema the statement ran as. This will make using the table for debugging purposes more difficult, e.g. if I have a "development" and a "production" schema, I would not know which schema the `SELECT`s that are being recorded are targeting.

Alternatively, we can store the statements after they have been name-resolved and re-serialized, but this I think is going to cause a performance hit.

**Comment by philip-stoev** [comment]
Item No 8. DDL and dataflow errors such as those:

```
materialize=> create table t2 (f1 integer, f1 integer);
ERROR:  column "f1" specified more than once
```

and:

```
materialize=> select f1/0 from t1;
ERROR:  Evaluation error: division by zero
```

are properly recorded in the table, but errors such as those:

```
materialize=> select f2 from t1;
ERROR:  column "f2" does not exist
```

and

```
materialize=> insert into t2 values (1);
ERROR:  unknown catalog item 't2'
```

and:
```
materialize=> select f1::timestamp from t1;
ERROR:  CAST does not support casting from integer to timestamp
```

are not. The common pattern seems that the errors not recorded happen during name resolution and type checking.

**Review by philip-stoev — CHANGES_REQUESTED** [review]
@umanwizard if you do not mind, I suggest the following testing:

- write an SLT test that would check for the proper recording of various situations, like the ones listed as Items
- consider running the entire CI by default with some ratio of recorded statements, e.g. 0.1 or 0.01. This would serve to exercise the code and possibly weed out any panics or other issues that may occur, without actually SELECTing from the newly-introduced tables.

File: src/adapter/src/coord.rs [hunk 14]
```diff
@@ -651,10 +650,44 @@ impl PendingRead {
...
+        statement_uuid.take()
+    }
+}
+
+impl Drop for ExecuteContextExtra {
+    fn drop(&mut self) {
+        let Self { statement_uuid } = &*self;
+        soft_assert_or_log!(
+            statement_uuid.is_none(),
+            "execute context dropped without being properly retired."
```
Comment (philip-stoev) [inline src/adapter/src/coord.rs:687]: This is triggered in Nightly if the feature is enabled by default:

(https://buildkite.com/materialize/nightlies/builds/3116#0189d44b-1a9c-4f0f-b46d-2a54697a3998)

**Comment by umanwizard** [comment]
@philip-stoev Thanks for the detailed report.

**Item 1**: I'll investigate.
**Item 2**: This is by design and matches the fact that e.g. `SET CLUSTER` doesn't error if you set it to something that could be a legal cluster name, even if it's not in fact a real cluster currently.
**Item 3**: That is by design. In the future we will probably the default to be above 0.
**Item 4**: In this case there are two statements involved: the outer statement (`EXECUTE q1(1234);`) and the inner statement that it references. Currently we are only logging execution of the outermost statement. Note that the outermost statement here does not have any parameters. If you want to observe parameterized statements you will need to use a postgres client library directly as it is not exposed to psql. I know this is confusing and we will probably change it soon but I want to leave dealing with inner statements out of this PR as it's already gotten huge.
**Item 5**: I think doing this might be quite difficult with our current architecture, but I'll take a look to check for sure.
**Item 6**: Similar to Item 4, these are not exposed to psql.
**Item 7**: We are going to store fully normalized/unambiguous statements as a follow-up.
**Item 8**: Will investigate.

To summarize: I will investigate items 1 and 8 as these may be unexpected bugs. Everything else is a planned follow-up or working as designed.

Please let me know if you don't agree.

**Comment by umanwizard** [comment]
*Item 1*: I can no longer repro this after rebasing.

**Comment by umanwizard** [comment]
*Item 8*: Ah, unfortunately this is working as designed (though I hadn't thought about this edge case) and is difficult to fix without conceptually changing how the entire thing works. The problem is that we aren't actually logging _statements_, but rather _executions of statements_. Thus, a statement that the user never attempted to execute will never be logged. When an error happens during _parsing_ or _planning_, it never gets to the execution stage and thus is never logged here.

I think we should fix this by creating separate tables to catch all those errors (called something like `mz_statement_planning_errors` and `mz_statement_parse_errors`), but IMO it's outside the scope of the present PR.

**Comment by benesch** [comment]
> called something like `mz_statement_planning_errors` and `mz_statement_parse_errors`

in the weeds thought: I'd unify those under `mz_statement_preparation_errors`

**Comment by umanwizard** [comment]
With my latest commit I cannot reproduce the `debug_assert` panic on nightly.

**Comment by umanwizard** [comment]
@philip-stoev This should be ready for another look from the QA side. I updated the default in mzcompose so this will always be running with 100% sampling in CI.

**Comment by philip-stoev** [comment]
@umanwizard this still need a dedicated test. Will you be able to write one or the QA team should do it?

**Comment by umanwizard** [comment]
We do have some dedicated tests, in Rust. If you think it will be useful I can add more in .slt/.td.

**Comment by umanwizard** [comment]
I've added a .td test which encodes the current behavior for the remaining issues that Philip found, which are "working as designed" (but for which we will add follow-ups later). This is in addition to the Rust tests we already have currently.

**Review by philip-stoev — APPROVED** [review]
